rfd/0035-aws-node-join.md

```diff
@@ -0,0 +1,294 @@
+   configured AWS Organization and/or Account.
+
+### tl;dr
+
+Configure auth server to allow nodes to join with the new aws join method:
+```yaml
+teleport:
+  nodename: auth
+
+auth_service:
+  enabled: yes
+  aws_join:
+    allow:
+    - organization: "o-1111111111"
+```
+
+Create ec2 instances in organization "o-1111111111" with a role with the
+following attached policy:
+```json
+{
+    "Version": "2012-10-17",
+    "Statement": [
+        {
+            "Effect": "Allow",
+            "Action": [
+                "organizations:DescribeOrganization"
+            ],
+            "Resource": "*"
+        }
+    ]
+}
+```
+
+Configure teleport nodes on ec2 instances:
+```yaml
+teleport:
+  auth_servers:
+    - auth
+  aws_token:
+    credential_source:
+      type: ec2
+    claims:
+    - organization
+
+ssh_service:
+  enabled: yes
+```
+
+Nodes will be authenticated with the new aws join method and be able to join
+the cluster.
+
+## Why
+
+This will make provisioning nodes on ec2 simpler and arguably more secure.
+There is no need to configure static tokens, or to manage the provisioning and
+deployment of dynamic tokens.
+
+## Details
+
+Much inspiration is taken from Vault's [IAM auth method](https://www.vaultproject.io/docs/auth/aws#iam-auth-method).
+
+### Authentication
+
+In place of a join token, nodes will present one or more signed AWS API
+requests to the auth server based on the configured claims:
+- `sts:GetCallerIdentity` for the "identity" claim and
+- `organizations:DescribeOrganization` for the "organization" claim
+
+These are signed HTTP requests which will be serialized and sent to the auth
+server over GRPC.
+
+The auth server will then:
+1. Check that the request URL is a valid AWS API endpoint.
+2. Send each request to the AWS API.
+  - AWS will check the signature and return an error if it is invalid, in which case the node will be rejected.
+3. Extract the Account from the `sts:GetCallerIdentity` response, if the request was provided.
+4. Extract the Organization from the `organizations:DescribeOrganization` response, if the request was provided.
+5. Check the Organization and Account against the configured deny rules. Reject the node if any match.
+6. Check the Organization and Account against the configured allow rules. Reject the node if none match.
+7. Possibly extra checks to prevent replay attacks, see the following section.
+
+By presenting signed AWS API requests, the node proves that it has access to
+AWS credentials. We rely on AWS to verify the signatures.
+
+The API responses prove the Organization and Account of the credentials
+used to sign the requests. We check that the request is a real AWS API
+endpoint, and rely on AWS to return the real identity for the signed
+requests.
+
+Nodes never share their AWS credentials, only signed requests.
+
+### Mitigating the Risk of Replay Attacks
+
+Signed API requests will only be held in memory and never logged or written to
```

Comment:
One of good ways to prevent replay attacks is to let teleport auth server to issue a random string of crypto-random generated bytes as a challenge and ask the node that wishes to join to include the challenge in the signed request with a TTL.

Reply:
GRPC streams could be an interesting help here:

* Node -> Creates a stream "Join stream"
* Sends a message -> Register Request
* Server responds -> Challenge bytes
* Node signs request and send the challenge response
* Server responds with an ephemeral join token for node to join or simply joins the node (depending on what's more convenient to implement)

Reply:
Oh that's pretty interesting, then the signed request would only be valid in the context of the original gRPC connection and replays are basically not a concern. Should be doable, I would probably implement like:
1. Node -> Creates bidirectional "Join stream"
2. Auth -> Sends challenge bytes
3. Node -> Sends signed AWS requests
4. Auth -> Checks that challenge bytes are included in signed headers, checks the signed requests with AWS, checks allow/deny rules, joins the node and sends keys

Reply:
The latest revision describes both the "EC2" and "IAM methods" for implementing this, with this challenge mechanism included in the "IAM" method